presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/OptimizeJoinsOverMapColumns.java

```diff
@@ -0,0 +1,456 @@
+                    joinNode.getDistributionType(),
+                    joinNode.getDynamicFilters());
+
+            if (isTopJoin) {
+                return transformJoin(newJoinNode, context.get());
+            }
+
+            return newJoinNode;
+        }
+
+        private boolean needsRewrite(JoinNode.Type joinType, List<VariableReferenceExpression> leftColumns, Set<VariableReferenceExpression> joinKeys)
+        {
+            if (joinType != JoinNode.Type.LEFT || !supportedJoinKeyTypes(joinKeys)) {
+                return false;
+            }
+
+//            Set<VariableReferenceExpression> leftMapColumns = extractMapColumns(leftColumns);
+//            return !leftMapColumns.isEmpty() && !leftMapColumns.containsAll(joinKeys);
+            // TODO: refine the condition to a less permissive one, e.g. taking into account size of payload not included in join keys
+            return !joinKeys.containsAll(leftColumns);
+        }
+
+        @Override
+        public PlanNode visitProject(ProjectNode projectNode, RewriteContext<JoinContext> context)
+        {
+            if (PlannerUtils.isScanFilterProject(projectNode)) {
+                return rewriteScanFilterProject(projectNode, context);
+            }
+
+            ProjectNode newProjectNode = (ProjectNode) context.defaultRewrite(projectNode, context.get());
+
+            if (projectNode != newProjectNode && !validateProjectAssignments(newProjectNode)) {
+                // some columns needed for the project were hidden by the rewrite: cancel rewrite
+                return projectNode;
+            }
+            return newProjectNode;
+        }
+
+        @Override
+        public PlanNode visitFilter(FilterNode filterNode, RewriteContext<JoinContext> context)
+        {
+            Set<VariableReferenceExpression> joinKeys = context.get().getJoinKeys();
+            List<VariableReferenceExpression> predicateVars = VariablesExtractor.extractAll(filterNode.getPredicate());
+
+            if (PlannerUtils.isScanFilterProject(filterNode)) {
+                return rewriteScanFilterProject(filterNode, context);
+            }
+
+            if (joinKeys.isEmpty() || joinKeys.containsAll(predicateVars)) {
+                return context.defaultRewrite(filterNode, context.get());
+            }
+
+            return filterNode;
+        }
+
+        @Override
+        public PlanNode visitTableScan(TableScanNode scanNode, RewriteContext<JoinContext> context)
+        {
+            return rewriteScanFilterProject(scanNode, context);
+        }
+
+        private PlanNode rewriteScanFilterProject(PlanNode planNode, RewriteContext<JoinContext> context)
+        {
+            Set<VariableReferenceExpression> joinKeys = context.get().getJoinKeys();
+
+            if (joinKeys.size() == 0 || context.get().getNumJoins() < 2) {
+                return planNode;
+            }
+
+            if (!ImmutableSet.copyOf(planNode.getOutputVariables()).containsAll(joinKeys)) {
```

Comment:
Why copyOf?

Reply:
getOutputVariables returns a list, this converts the list to a set to check for containment